File: src/DataScience/R/builder.R
text_builder <- function() {
  parts <- character()
  list(
    reset = function() { parts <<- character() },
    add_title = function(title) { parts <<- c(parts, paste0("# ", title)) },
    add_section = function(heading, body) { parts <<- c(parts, paste0("## ", heading), body) },
    build = function() paste(parts, collapse = "\n")
  )
}

html_builder <- function() {
  parts <- character()
  list(
    reset = function() { parts <<- character() },
    add_title = function(title) { parts <<- c(parts, paste0("<h1>", title, "</h1>")) },
    add_section = function(heading, body) {
      parts <<- c(parts, paste0("<h2>", heading, "</h2>"), paste0("<p>", body, "</p>"))
    },
    build = function() paste(parts, collapse = "")
  )
}

build_availability_report <- function(builder) {
  builder$reset()
  builder$add_title("Service status")
  builder$add_section("Availability", "99.95%")
  builder$build()
}

cat(build_availability_report(text_builder()), "\n---\n", sep = "")
cat(build_availability_report(html_builder()), "\n", sep = "")

File: src/DataScience/Octave/builder.m
function builder()
  disp(build_availability_report(text_builder()));
  disp('---');
  disp(build_availability_report(html_builder()));
end

function b = text_builder()
  parts = {};
  b.reset = @reset;
  b.add_title = @add_title;
  b.add_section = @add_section;
  b.build = @build;

  function reset()
    parts = {};
  end
  function add_title(title)
    parts{end + 1} = ['# ' title];
  end
  function add_section(heading, body)
    parts{end + 1} = ['## ' heading];
    parts{end + 1} = body;
  end
  function result = build()
    result = strjoin(parts, sprintf('\n'));
  end
end

function b = html_builder()
  parts = {};
  b.reset = @reset;
  b.add_title = @add_title;
  b.add_section = @add_section;
  b.build = @build;

  function reset()
    parts = {};
  end
  function add_title(title)
    parts{end + 1} = ['<h1>' title '</h1>'];
  end
  function add_section(heading, body)
    parts{end + 1} = ['<h2>' heading '</h2>'];
    parts{end + 1} = ['<p>' body '</p>'];
  end
  function result = build()
    result = strjoin(parts, '');
  end
end

function result = build_availability_report(b)
  b.reset();
  b.add_title('Service status');
  b.add_section('Availability', '99.95%');
  result = b.build();
end
